src/lib.rs

```diff
@@ -1,9 +1,361 @@
+    /// message, otherwise this side will wait for the other side to send the join message.
+    ///
+    /// # Arguments
+    ///
+    /// * source - An identifier for the source of the connection. Only
+    ///   currently used for logging and typically is a
+    ///   [`std::net::SocketAddr`]
+    /// * stream - The stream + sink of messages representing the connection
+    /// * direction - Which side of the connection `stream` is
+    ///
+    /// # Sharing
+    ///
+    /// Once the connection is established and we know the other ends peer ID
+    /// the repo will query all the documents it knows about using `Store::list` and
+    /// then send sync messages for every document for which
+    /// `Share::should_share` returns `ShareDecision::Share`
+    pub async fn connect<RecvErr, SendErr, Source>(
+        self,
+        source: Source,
+        stream: Store,
+        direction: ConnDirection,
+    ) -> Result<(), error::RunConnection<SendErr, RecvErr>>
+    where
+        RecvErr: std::error::Error,
+        SendErr: std::error::Error,
+        Store: Sink<Message, Error = SendErr> + Stream<Item = Result<Message, RecvErr>>,
+        Source: std::fmt::Debug + Eq,
+    {
+        todo!()
+    }
+
+    /// Connect a tokio io object
+    ///
+    /// This implements a simple length prefixed framing protocol and is intended for use with
+    /// stream oriented transports
+    #[cfg(feature = "tokio")]
+    pub async fn connect_tokio_io<Io, Source>(
+        self,
+        source: Source,
+        io: Io,
+        direction: ConnDirection,
+    ) -> Result<(), error::RunConnection<error::CodecError, error::CodecError>>
+    where
+        Io: tokio::io::AsyncRead + tokio::io::AsyncWrite,
+    {
+        todo!()
+    }
+
+    /// Attempt to load the document given by `id` from storage
+    ///
+    /// If the document is not found in storage then `Ok(None)` is returned.
+    ///
+    /// Note that this _does not_ attempt to fetch the document from the
+    /// network.
+    pub async fn load(&self, id: DocumentId) -> Result<Option<DocHandle>, Store::Error> {
+        todo!()
+    }
+
+    /// Request a document from a peer in the network
+    ///
+    /// This will connect to the peers specified by `peers` and run the sync
+    /// protocol for the given document ID. Once the sync protocol has run (or
+    /// not as the case may be) this function will return the document handle.
+    ///
+    /// This can be used both to fetch a document you don't have and to request
+    /// updates you think another peer has but which you haven't been sent yet.
+    ///
+    /// # Arguments
+    ///
+    /// * id - The ID of the document to request
+    /// * peer - The peer(s) to request the document from.
+    pub async fn request(
+        &self,
+        id: DocumentId,
+        peers: PeerSpec,
+    ) -> Result<Option<DocHandle>, error::Request> {
+        todo!()
+    }
```

Comment:
I assume this is the mirror of `announce`? 

I'd say when a peer connects we should simply start syncing all documents we know of as a default, and perhaps later add APIs giving the user more control. 

Announcing the IDs only would be a way to perhaps optimize that(you'd trade smaller initial messages for more back and forth).

Reply:
Not exactly a mirror. The point of this method is that it encapsulates the idea that you know a document ID, but you don't have the document in storage (maybe you already tried `Repo::load` and got back a `None`) so now you know you need to ask the network. The implementation may in fact not do anything as maybe it already has up to date sync channels with the specified peers, so it can just return immediately. On the other hand, the `Repo` may have just started up and is consequently still connecting to the specified peers (not "specified" is loose here, the specification can just be "any peers") and so the method provides a way to wait for the sync to have happened.

This is also a hook for in future doing more complex optimizations as you suggest, but the point right now is to evoke that this is for getting things from the network vs `load`, which gets them from storage.